### ISSUE TITLE
Running jsReload causes sub-modals to multiply in modal->set()

### ISSUE THREAD

**Issue by pkly** [atk4/ui#681]
Test scenario

```
$modal = $this->_app->add(['Modal', 'title' => 'Test']);

$this->_app->add(['Button', 'Whatever'])->on('click', $modal->show());

$modal->set(function($v) {
    $v->add(['Modal', 'title' => 'Here!']);
    $v->add(['Modal', 'title' => 'Another!']);

    $v->add(['Button', 'Reload me'])->on('click', $v->jsReload());
});
```

Expected behavior: 
Clicking the "Reload me" button reloads the view.

Actual experience:
Clicking the "Reload me" button reloads the view, but causes the two modals created inside of it to get a new copy (each).

First `->set()` [Image](https://i.imgur.com/0RdXD0g.png)
After `->jsReload()` [Image](https://i.imgur.com/IoWCxwQ.png) (notice the duplicate IDs)

This is a rather nasty bug.

In this scenario running `->jsReload()` on the modal also makes it kinda, disappear and die.

What's even the solution to something like this? I used to refresh a modal by parts, but that actually caused it to lag badly because of the numer of requests required.

A suggested solution would be to modify ApiService in a way so it can pre-process the html response.
in `ApiService.js` on line **75** modify the value of response.html, or let it be read by another service/plugin beforehand.

I believe running something this would fix this issue.
I might play around with the idea at a later time, since I wanted to add pre-actions.

```
let modalIDs = [];
$(response.html).find(".modal[id]").each((i, e) => {
    modalIDs.push('#' + $(e).attr('id'));
});

if (modalIDs.length) 
    $('body').find(modalIDs.join(', ')).remove();

// old line 75 in ApiService.js
result = [..]
```

**Comment by ibelar** [comment]
@pkly  - This is an issue we are having with Fomantic(Semantic) modal. 
When you reload view in code above,  this will issue the jQuery command $('modal').modal() for Formantic and this will create the modal view whether it exist or not and create duplicate.

Hacking modalService will not help either because when modalService is invoke, modals are already created.

Recommendation would be to avoid adding Modal to your reload view but if you need to do so, here is a quick workaround using your code above:

```
$modal = $this->_app->add(['Modal', 'title' => 'Test']);

$this->_app->add(['Button', 'Whatever'])->on('click', $modal->show());

$modal->set(function($v) {
   $isLoaded = $_GET['loaded'] ? $_GET['loaded'] : null;
    if (!$isLoaded) {
        $v->add(['Modal', 'title' => 'Here!']);
        $v->add(['Modal', 'title' => 'Another!']);
    }

    $v->add(['Button', 'Reload me'])->on('click', $v->jsReload(['loaded' => 1]));
});
```

This will make sure Modal in your reload view are created only once.

Hope this help!

**Comment by pkly** [comment]
@ibelar wouldn't the solution I proposed work though? It modifies the ApiService class which reloads the html.
Like, you're just replacing html of the main element, which will now contain the same modals again.
Basically you're re-adding the same elements (modals, in this case), without really realizing it.
Since this code is in the atk js application, it doesn't seem like an error on Fomantic's side?

Since you're essentially duplicating IDs, I'm unsure of how this should be even handled, since that's not allowed behavior.
Does Fomantic actually handle element reloading, or is that handled purely by atk? I believed it's the latter.

Modifying ApiService will assure that the modals are removed before their creating call is ran, thus causing a "real" refresh of any elements in the modals, and the modals themselves.

**Comment by pkly** [comment]
Besides, the solution you've proposed doesn't even do what a reload should do.
You know, reload things.

Granted, this is probably not the intended usage of jsReload that you guys had in mind, but if it's there it should work properly in all situations, shouldn't it?

**Comment by ibelar** [comment]
@pkly - My apologies. I have miss read your first comment. I thought you where referring to modalService, not apiService. Actually your proposal solution looks very good. 

I might change it a bit in order to prevent from erasing a custom view using the class name "modal". I believe if we look for .ui.modal together instead of "modal" only and find existing one in ".ui.dimmer.modals.page" instead of body we should by safe.

Thanks, definitely worth trying!

**Comment by ibelar** [comment]
fix by #684
